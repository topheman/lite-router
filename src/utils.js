/**
 *
 * lite-router
 *
 * @author Christophe Rosset - @topheman
 *
 */

/**
 * Patterns accepted:
 * - "*": match all
 * - RegExp
 * - /post/:param1/edit/:param2 -> will match this pattern and return the matched params in a params object
 * - a string
 *
 * This method will attempt to return a matcher, according to the pattern returned
 * This matcher has the signature function(pathname) {...} will return false or the matched params
 * It takes a pattern in param and return a function that takes a pathname in param and will return:
 *  - false if no match
 *  - an object with potential matched params
 * @param {String|RegExp} pattern
 * @returns {Function} (pathname) => false|{paramsMatched}
 */
export const compilePattern = (pattern) => {
  // 1) if pattern is a regexp, return a matcher using this regexp
  if (pattern instanceof RegExp) {
    return (pathname) => {
      if (pathname.match(pattern) !== null) {
        return {};
      }
      return false;
    };
  }

  // 2) if pattern is a "match all", return a matcher that always matches
  if (pattern === '*') {
    // this matcher matches any pathname
    return () => ({});
  }

  // 3) try to create a matcher for a pattern with params in it

  const paramNamesMatcher = /:([a-zA-Z_$][a-zA-Z0-9_$]*)/g;
  let paramNames = pattern.match(paramNamesMatcher);
  // console.log('paramNames', paramNames);

  if (paramNames !== null) {
    paramNames = paramNames.map(match => match.slice(1));// remove the ":" to get the exact paramNames
    const paramNamesRegexpSource = pattern.replace(/:([^\/]+)/g, '([^/?#]+)');// replace ":paramName" by the regexp that will match this param
    return (pathname) => {
      const matches = pathname.match(new RegExp(`^${paramNamesRegexpSource}$`));
      let matchedParams;
      if (matches !== null) {
        matchedParams = matches.reduce((result, match, index) => {
          if (index > 0) {
            result[paramNames[index - 1]] = match;// eslint-disable-line no-param-reassign
          }
          return result;
        }, {});
        return matchedParams;
      }
      return false;
    };
  }

  // 4) if nothing of the above applied, simply return a matcher that matched a string
  return (pathname) => (pathname === pattern ? {} : false);
};

/**
 * Returns an array of routes containing matcher & handler
 * Throws error if any param missing
 * @param routes
 * @throws
 */
export const normalizeRoutes = (routes) => routes.map(route => {
  if (typeof route.pattern === 'undefined' || typeof route.handler === 'undefined') {
    throw new Error('Missing pattern or handler attribute');
  }
  if (typeof route.resolve !== 'undefined' && typeof route.resolve.then === 'undefined') {
    throw new Error('resolve param only accepts promises');
  }
  return {
    matcher: compilePattern(route.pattern),
    handler: route.handler,
    resolve: route.resolve
  };
});

/**
 * Returns a function that will match a pathname (from location.pathname) to a mount (a route with handler infos)
 * @param routes
 */
export const compileMatchMount = (routes) => (pathname) => { // eslint-disable-line arrow-body-style
                                                             // match the location.pathname to one of the routes and extract the related mounting infos (handler, resolve ...)
  return routes
    .reduce((result, route) => {
      const params = route.matcher(pathname);// a matcher returns false if no match or an object with potentials params matched for the route
      if (params && result.length === 0) { // once we get a match, no more matching
        result.push({
          handler: route.handler,
          resolve: route.resolve,
          params
        });
      }
      return result;
    }, [])
    .reduce((result, matchedMount) => result || matchedMount, null);// 1) always reducing to the first match if multiple ones 2) if no match, reduce to null
};
