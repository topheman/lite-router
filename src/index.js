import { add } from './utils';

const addTen = (n) => add(n, 10);

export default addTen;