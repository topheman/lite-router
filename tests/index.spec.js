/* eslint-disable no-unused-expressions */

import { compilePattern, normalizeRoutes, compileMatchMount } from '../src/utils';

describe('micro-router', () => {
  describe('compilePattern', () => {
    describe('Match all pattern: *', () => {
      const matcher = compilePattern('*');
      it('should match /toto', () => {
        expect(matcher('/toto')).toEqual({});
      });
      it('should match anything', () => {
        expect(matcher('anything')).toEqual({});
      });
    });
    describe('RegExp pattern: /^\\/posts\\/(\\w+)\\/(\\w+)\\/edit$/', () => {
      const matcher = compilePattern(/^\/posts\/(\w+)\/(\w+)\/edit$/);
      it('should match /posts/foo/bar/edit', () => {
        expect(matcher('/posts/foo/bar/edit')).toEqual({});
      });
      it('should NOT match /posts/foo/bar/editNO', () => {
        expect(matcher('/posts/foo/bar/editNO')).toBe(false);
      });
      it('should NOT match /posts/foo/bar/edit/NO', () => {
        expect(matcher('/posts/foo/bar/editNO')).toBe(false);
      });
      it('should NOT match /posts/foo/bar', () => {
        expect(matcher('/posts/foo/bar')).toBe(false);
      });
    });
    describe('Params pattern: /repos/:owner/:name/edit', () => {
      const matcher = compilePattern('/repos/:owner/:name/edit');
      it('should match /repos/topheman/rsjs-experiments/edit', () => {
        expect(matcher('/repos/topheman/rsjs-experiments/edit')).toBeTruthy();
      });
      it('should match the params from /repos/topheman/rsjs-experiments/edit', () => {
        expect(matcher('/repos/topheman/rxjs-experiments/edit')).toEqual({
          owner: 'topheman',
          name: 'rxjs-experiments'
        });
      });
      it('should NOT match /repo/reactjs/redux', () => {
        expect(matcher('/repo/reactjs/redux')).toBe(false);
      });
      it('should NOT match /repo/reactjs/redux/display', () => {
        expect(matcher('/repo/reactjs/redux/display')).toBe(false);
      });
      it('should NOT match /repo/reactjs', () => {
        expect(matcher('/repo/reactjs')).toBe(false);
      });
      it('should NOT match /repo', () => {
        expect(matcher('/repo/reactjs')).toBe(false);
      });
    });
    describe('Plain string patter: /hello/world', () => {
      const matcher = compilePattern('/hello/world');
      it('should match /hello/world', () => {
        expect(matcher('/hello/world')).toEqual({});
      });
      it('should NOT match /hello/world/NOT', () => {
        expect(matcher('/hello/world/NOT')).toBe(false);
      });
      it('should NOT match /hello', () => {
        expect(matcher('/hello')).toBe(false);
      });
    });
  });
  describe('normalizeRoutes', () => {
    describe('sanity checks', () => {
      it('should NOT throw when pattern and handler attribute are present', () => {
        expect(() => normalizeRoutes([
          { pattern: 'foo', handler: () => ({}) }
        ])).not.toThrowError();
      });
      it('should throw when handler attribute is missing', () => {
        expect(() => normalizeRoutes([
          { pattern: 'foo' }
        ])).toThrowError();
      });
      it('should throw when pattern attribute is missing', () => {
        expect(() => normalizeRoutes([
          { handler: () => ({}) }
        ])).toThrowError();
      });
      it('should throw if resolve attribute is not a promise', () => {
        expect(() => normalizeRoutes([
          { pattern: 'foo', handler: () => ({}), resolve: {} }
        ])).toThrowError();
      });
      it('should NOT throw if resolve attribute is a promise', () => {
        expect(() => normalizeRoutes([
          { pattern: 'foo', handler: () => ({}), resolve: { then: () => ({}) } }
        ])).not.toThrowError();
      });
    });
    describe('mapping - normalizeRoutes(routes)', () => {
      const routes = [
        { pattern: '/hello/world', handler: () => ({}) },
        { pattern: /^\/hie\/universe/, handler: () => ({}) },
        { pattern: '/hello/:name', handler: () => ({}) },
        { pattern: '*', handler: () => ({}) }
      ];
      it('should return a array', () => {
        expect(normalizeRoutes(routes)).toHaveLength(4);
      });
      it('should return a populated array', () => {
        const normalizedRoutes = normalizeRoutes(routes);
        expect(normalizedRoutes[0]).toMatchSnapshot();
        expect(normalizedRoutes[1]).toMatchSnapshot();
        expect(normalizedRoutes[2]).toMatchSnapshot();
        expect(normalizedRoutes[3]).toMatchSnapshot();
      });
    });
  });
  describe('compileMatchMount', () => {
    const routes = [
      { pattern: '/hello/world', handler: () => ({}) },
      { pattern: /^\/hie\/universe/, handler: () => ({}) },
      { pattern: '/hello/:name', handler: () => ({}) },
      { pattern: '*', handler: () => ({}) }
    ];
    const matchMount = compileMatchMount(normalizeRoutes(routes));
    it('should match /hello/world', () => {
      expect(matchMount('/hello/world')).toMatchSnapshot();
      expect(matchMount('/hello/world').params).toEqual({});
    });
    it('should match /hie/universe', () => {
      expect(matchMount('/hie/universe')).toMatchSnapshot();
      expect(matchMount('/hie/universe').params).toEqual({});
    });
    it('should match /hello/topheman', () => {
      expect(matchMount('/hello/topheman')).toMatchSnapshot();
      expect(matchMount('/hello/topheman').params).toEqual({ name: 'topheman' });
    });
    it('should match /whatever', () => {
      expect(matchMount('/whatever')).toMatchSnapshot();
      expect(matchMount('/whatever').params).toEqual({});
    });
  });
});
