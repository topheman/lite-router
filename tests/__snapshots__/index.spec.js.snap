// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`micro-router compileMatchMount should match /hello/topheman 1`] = `
Object {
  "handler": [Function],
  "params": Object {
    "name": "topheman",
  },
  "resolve": undefined,
}
`;

exports[`micro-router compileMatchMount should match /hello/world 1`] = `
Object {
  "handler": [Function],
  "params": Object {},
  "resolve": undefined,
}
`;

exports[`micro-router compileMatchMount should match /hie/universe 1`] = `
Object {
  "handler": [Function],
  "params": Object {},
  "resolve": undefined,
}
`;

exports[`micro-router compileMatchMount should match /whatever 1`] = `
Object {
  "handler": [Function],
  "params": Object {},
  "resolve": undefined,
}
`;

exports[`micro-router normalizeRoutes mapping - normalizeRoutes(routes) should return a populated array 1`] = `
Object {
  "handler": [Function],
  "matcher": [Function],
  "resolve": undefined,
}
`;

exports[`micro-router normalizeRoutes mapping - normalizeRoutes(routes) should return a populated array 2`] = `
Object {
  "handler": [Function],
  "matcher": [Function],
  "resolve": undefined,
}
`;

exports[`micro-router normalizeRoutes mapping - normalizeRoutes(routes) should return a populated array 3`] = `
Object {
  "handler": [Function],
  "matcher": [Function],
  "resolve": undefined,
}
`;

exports[`micro-router normalizeRoutes mapping - normalizeRoutes(routes) should return a populated array 4`] = `
Object {
  "handler": [Function],
  "matcher": [Function],
  "resolve": undefined,
}
`;
